src/nativeAssetManager.js

```diff
@@ -27,10 +29,134 @@ const NATIVE_KEYS = {
   salePrice: 'hb_native_saleprice',
 };
 
+// Asset type mapping as per Native IAB spec 1.2
+// https://www.iab.com/wp-content/uploads/2017/04/OpenRTB-Native-Ads-Specification-Draft_1.2_2017-04.pdf#page=40
+const assetTypeMapping = {
+  'image': {
+    1: 'icon',
+    3: 'image'
+  },
+  'data': {
+    1: 'sponsoredBy',
+    2: 'body',
+    3: 'rating',
+    4: 'likes',
+    5: 'downloads',
+    6: 'price',
+    7: 'salePrice',
+    8: 'phone',
+    9: 'address',
+    10: 'body2',
+    11: 'displayUrl',
+    12: 'cta',
+  }
+}
+
+const DEFAULT_CACHE_HOST = 'prebid.adnxs.com';
+const DEFAULT_CACHE_PATH = '/pbc/v1/cache';
+
 export function newNativeAssetManager(win) {
   let callback;
   let errorCountEscapeHatch = 0;
 
+  function getCacheEndpoint(cacheHost, cachePath) {
+    let host = (typeof cacheHost === 'undefined' || cacheHost === "") ? DEFAULT_CACHE_HOST : cacheHost;
+    let path = (typeof cachePath === 'undefined' || cachePath === "") ? DEFAULT_CACHE_PATH : cachePath;
+  
+    return `https://${host}${path}`;
+  }
+  
+  function parseResponse(response) {
+    let bidObject;
+    try {
+      bidObject = JSON.parse(response);
+    } catch (error) {
+      console.log(`Error parsing response from cache host: ${error}`);
+    }
+    return bidObject;
+  }
+  
+  function transformToPrebidKeys(adMarkup) {
+    let assets = [];
+    let clicktrackers;
+    let assetsFromMarkup = adMarkup.assets;
+    assetsFromMarkup.forEach((asset) => {
+      if (asset.img) {
+        if (assetTypeMapping['image'][asset.img.type]) {
+          assets.push({
+            'key' : assetTypeMapping['image'][asset.img.type],
+            'value' : asset.img.url
+          })
+        } else {
+          console.log('ERROR: Invalid image type for image asset');
+        }
+      } else if (asset.data) {
+        if (assetTypeMapping['data'][asset.data.type]) {
+          assets.push({
+            'key' : assetTypeMapping['data'][asset.data.type],
+            'value' : asset.data.value
+          })
+        } else {
+          console.log('ERROR: Invalid data type for data asset');
+        }
+      } else if (asset.title) {
+        assets.push({
+          'key' : 'title',
+          'value' : asset.title.text
+        })
+      } 
+    })
+
+    if (adMarkup.link) {
+      if (adMarkup.link.clicktrackers) {
+        clicktrackers = adMarkup.link.clicktrackers;
+      }
+      assets.push({
+        'key' : 'clickUrl',
+        'value' : adMarkup.link.url
+      })
+    }
+
+    return {
+      assets,
+      clicktrackers,
+      'imptrackers' : adMarkup.imptrackers
+    }
+  }
+
+  function requestAssetsFromCache(tagData) {
+    let ajaxCallback = function(response) {
+      let bidResponse = parseResponse(response);
+      if (bidResponse.adm) {
```

Comment:
`bidResponse` will be undefined if the JSON.parse(response) fails, in which case this line fails.